Firestore/core/src/util/thread_safe_memoizer.h

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_UTIL_THREAD_SAFE_MEMOIZER_H_
+#define FIRESTORE_CORE_SRC_UTIL_THREAD_SAFE_MEMOIZER_H_
+
+#include <mutex>  // NOLINT(build/c++11)
+#include <vector>
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * Stores a memoized value in a manner that is safe to be shared between
+ * multiple threads.
+ */
+template <typename T>
+class ThreadSafeMemoizer {
+ public:
+  ~ThreadSafeMemoizer() {
+    // Call `std::call_once` in order to synchronize with the "active"
+    // invocation of `memoize()`. Without this synchronization, there is a data
+    // race between this destructor, which "reads" `memoized_value_` to destroy
+    // it, and the write to `memoized_value_` done by the "active" invocation of
+    // `memoize()`.
+    std::call_once(once_, [&]() {});
+  }
+
```

Comment:
Explicitly mark the copy and move constructors and assignment operators as `= delete`. That way it's "obvious" that those operations are not supported. It also means you don't need to test those operations ;)